src/runtime_src/core/common/xrt_xclbin.cpp

```diff
@@ -74,3 +79,375 @@ xrtXclbinUUID(xclDeviceHandle dhdl, xuid_t out)
     return -1;
   }
 }
+
+namespace xrt {
+
+// class xclbin_impl - class for xclbin objects
+//
+// Life time of xclbin are managed through shared pointers.
+// A buffer is freed when last references is released.
+class xclbin_impl
+{
+protected:
+  std::vector<char> m_axlf;
+  const axlf* m_top;
+
+  void init_axlf_handle()
+  {
+    const axlf* tmp = reinterpret_cast<const axlf*>(m_axlf.data());
+    if (strncmp(tmp->m_magic, "xclbin2", 7)) // Future: Do not hardcode "xclbin2"
+      throw std::runtime_error("Invalid xclbin");
+    m_top = tmp;
+  }
+
+public:
+  xclbin_impl() : m_top(nullptr) {}
+
+  xclbin_impl(const std::vector<char>& data)
+  {
+    m_axlf = data;
+    // Set pointer of type axlf*, and check magic string
+    init_axlf_handle();
+  }
+
+  xclbin_impl(const std::string& filename)
+  {
+    if (filename.empty())
+      throw std::runtime_error("No XCLBIN specified");
+
+    // Load the file into data
+    std::ifstream stream(filename);
+    stream.seekg(0, stream.end);
+    size_t size = stream.tellg();
+    stream.seekg(0, stream.beg);
+    m_axlf.resize(size);
+    stream.read(m_axlf.data(), size);
+
+    // Set pointer of type axlf*, and check magic string
+    init_axlf_handle();
+  }
+
+  bool
+  empty() const
+  {
+    return m_axlf.size() == 0 || m_top == nullptr;
+  }
+
+  void
+  not_empty_or_error() const
+  {
+    if (empty())
+      throw xrt_core::error(-EINVAL, "Invalid XCLBIN data");
+  }
```

Comment:
Not sure 131-141 are needed?

Reply:
Due to implementation shared above these functions are leveraged. They cause exception to be thrown when xclbin methods are called, if the xclbin is empty or invalid.

Reply:
If you remove the emptyness, xclbin_impl will always be valid, right?

Reply:
removed